### ISSUE TITLE
[Ballerina Shell] Global variable incorrect behavior in closures

### ISSUE THREAD

**Issue by kdsuneraavinash** [ballerina-platform/ballerina-lang#28036]
**Description:**
Assignments done to global variables in closures are not be reflected after the execution. The changes are visible only for the scope belonging to the snippet where the closure was defined. The reason is due to closure holding the previous snippet global variable reference.

**Steps to reproduce:**
In REPL,
```ballerina
=$ int x = 10
=$ var f = function () { x = 12; }
=$ f()
=$ x   // <- this should output 12 but will output 10 instead
```

This is the same for class methods or any variable holding a function reference. The global variables inside them are the wrong references.

**Comment by kdsuneraavinash** [comment]
This can be fixed by using a compiler phase for ballerina shell.
The cause of this issue is that the variables in closures point to the old reference. Since global variables are redefined at each step, the references in closure become invalid (or rather point to the same old reference) when the next snippet is evaluated.

One approach to fix this would be to remove global variable assignments/use and replace them directly with MEMORIZE and REMEMBER calls. So following code,
```ballerina
var f = function () {
    i += 1; // i is a global variable
}
```
would be converted to,
```ballerina
var f = function () {
   MEMORIZE("i", REMEMBER("i") + 1);  // i is a global variable
}
```

Since this would have to run after desugaring phase, (otherwise it may be required to desugar some code segments anyways), a compiler phase might be required.

This closed PR implements a similar approach.
https://github.com/ballerina-platform/ballerina-lang/pull/28458

**Comment by DamithaSenevirathne** [comment]
@kdsuneraavinash can you describe the reason why the mentioned PR/approach https://github.com/ballerina-platform/ballerina-lang/pull/28458 is closed without merging it with the master ? Any possible improvements to the above implementation ?

**Comment by kdsuneraavinash** [comment]
The reason it was closed without merging was that it was kind of a big change at the time (because it was implemented as a new compiler phase) I only implemented that PR as a reference implementation and by now I think some things in that PR may be broken.

**Comment by mohanvive** [comment]
We have failed to provide proper estimate for this. Hence, we have to add more points for this.

**Comment by mohanvive** [comment]
We have spent a few cycles to resolve this issue with the existing design of the Ballerina Shell. But we figured it was not possible to fix these issues entirely with the existing Shell architecture. There are a few options we could consider to completely fix this issue. 

We should discuss and come up with a plan as the next step.

1) Introduce another compiler phase that runs only for the Shell. Sunera did an implementation for this some time back. https://github.com/ballerina-platform/ballerina-lang/pull/28458 . But IMO, it is not the best approach
2) Remove the temporary memory storage used at Shell and redesign the Shell to act the simple client, which simply wraps user inputs and performs ballerina execution each user input.
3) Redesign the Shell to use the Semantic model instead of a syntax tree. This requires some improvement in the compiler to support semantic model execution.
